prov/shm/src/smr_rma.c

```diff
@@ -160,9 +163,30 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
 		}
 		resp = ofi_cirque_tail(smr_resp_queue(ep->region));
 		pend = freestack_pop(ep->pend_fs);
-		smr_format_iov(cmd, smr_peer_addr(ep->region)[peer_id].addr,
-			       iov, iov_count, total_len, op, 0, data,
-			       op_flags, context, ep->region, resp, pend);
+		if (ep->region->cma_cap) {
+			smr_format_iov(cmd, smr_peer_addr(ep->region)[peer_id].addr,
+				       iov, iov_count, total_len, op, 0, data,
+				       op_flags, context, ep->region, resp, pend);
+		} else {
+			/*
+			 * mmap protocol
+			 * Once SAR protocol gets merged, we need set a threshold
+			 * for switching from SAR to mmap
+			 */
+			msg_id = (ep->msg_id)++;
+			ret = smr_iov_mmap_copy_in(ep, peer_smr, iov,
+						   iov_count, total_len, op, msg_id,
+						   &map_name, &shm_ptr);
+			if (ret) {
+				freestack_push(ep->pend_fs, pend);
```

Comment:
Why is the stack push not part of the common error path?

Reply:
This is because only format_mmap (combined copy_in and original format_mmap) call could fail. format_iov call just does bunch of assignments.